### ISSUE TITLE
Extract Class

### ISSUE THREAD

**Issue by avik-so** [nicoespeon/abracadabra#45]
## Is this request related to a problem? Please describe.

Often when I'm refactoring, I find a hidden class. A collection of functions and properties that would be better in a new class or in another existing class.

## Describe the solution you'd like

I'd like to be able to select a function or more, maybe even some variables, and extract it out to a new class.
I'd also like for all existing calls to those functions to now be prepended with the new class.

## Additional context

For example, I'd like to turn this:

```ts
class Foo () {
  barSize: int = 30;
  fooString: string = "baz";

  doFoo () {
    console.log(fooString);
  }

  doBar() {
    return barSize /2;
  }

  doubleBar() {
    return barSize *2;
  }
}

class Baz() {
  Foo.doBar();
  Foo.doFoo();
}
```

into this:

```ts
class Foo () {
  fooString: string = "baz";

  doFoo () {
    console.log(fooString);
  }
}

class Bar() {
  barSize: int = 30;
  
  doBar() {
    return barSize /2;
  } 

  doubleBar() {
    return barSize *2;
  }
}

class Baz() {
  Bar.doBar();
  Foo.doFoo();
}
```

Being able to then move it into a new file would be a nice bonus.

**Comment by nicoespeon** [comment]
Oh, thanks for suggesting this one @avik-so 👍 

I see exactly what you mean. Indeed, I'd like to have this one done.

In practice, I think classes are generally located in different files. So I think that first, I'll need to figure out how to resolve the import-export relations across files. That's what prevented me to implement class-level refactorings until now. I'll create an issue for that.

**Comment by avik-so** [comment]
Webstorm has the "move" feature.  But it doesn't handle classes well.

**Comment by nicoespeon** [comment]
After giving a second thought about it, I think it's easier than I thought.

We don't actually need to index the project to do this refactoring, as I thought when I created https://github.com/nicoespeon/abracadabra/issues/46

Actually, all the information are contained in one file. Then, it's relatively easy to put the extracted class into another file and import it.

I will consider implementing this feature sooner than I first thought.

**Comment by justerest** [comment]
Hi, @nicoespeon !

I want to provide PR for Extract Class refactoring.

I recently made an own extension for class extraction [https://github.com/justerest/extract-class](https://github.com/justerest/extract-class) (not published to marketplace).

It's different from source proposal of this issue. No calls delegation from source class clients to extracted class. But I think it's optional. And it's has dependency [ts-morph](https://github.com/dsherret/ts-morph) to work with AST.

Can you review my project to shure that behavior are expected for Abracadabra project?

You can build and run or view refactoring examples at test file [https://github.com/justerest/extract-class/blob/master/src/ClassRefactor.test.ts](https://github.com/justerest/extract-class/blob/master/src/ClassRefactor.test.ts)

**Comment by nicoespeon** [comment]
Hey @justerest 👋 

I had a look at the tests examples and that looks very good. I'm impressed that you tackled many scenarios already and I think it can be quite handy!

That definitely matches what I would have in mind for Abracadabra. The goal would be to create a new class from an existing one, to split the responsibilities.

## Things to consider

I guess there would be a few things to determine while implementing it:
- Take the Selection (e.g. cursor position) into account to determine the source class to extract
- Provide a Quick Fix so the refactoring shows up when it can be executed
- Find a way to "ask" the user which methods/properties to extract (I'm not sure if you're doing that already) (although, we could start simple here)

Most of these are naturally handled in Abracadabra (e.g. you'll have the `selection` and a way to detect if it's "inside" the node).

## Getting started

If we port that refactoring, I can help you adapt to the expected I/O: https://github.com/nicoespeon/abracadabra/blob/1e4fd6cdb876746626c6dd546e2d1ab5af5c41be/src/types.ts#L14-L29

Using the generator would create the main frame: https://github.com/nicoespeon/abracadabra/blob/master/CONTRIBUTING.md#create-a-new-refactoring

## Regarding ts-morph, it will work fine

Abracadabra is using Babel to parse, traverse and transform the AST in general. But in practice, it doesn't really matter. Each refactoring can be implemented as we like. It receives an `Editor` that it can use to read the code and write the updated version. I think we should be fine adding ts-morph and try a different approach with this refactoring. Maybe it will inspire improvements for others!

Here's an example of a refactoring that's delegating the work to a React codemod: https://github.com/nicoespeon/abracadabra/blob/1e4fd6cdb876746626c6dd546e2d1ab5af5c41be/src/refactorings/react/convert-to-pure-component/convert-to-pure-component.ts

As you can see, this one doesn't use Babel to transform the code, but jscodeshift.

---

Therefore, thank you for proposing this. It has been in the backlog for a long time. It's not an easy refactoring, although it's really do-able (we've implemented other non-trivial refactorings here). So it takes time to do a good job.

You have a very nice implementation. I like how your tests are readable. 

If you want, I can help you pushing this. Open a draft PR, play around and ping me whenever you feel blocked 👍

**Comment by nicoespeon** [comment]
@avik-so I'm glad to tell you that, after so much time, this was implemented thanks to @justerest 🎉 

If I take your example:

```ts
class Foo {
  barSize: number = 30;
  fooString: string = "baz";

  doFoo () {
    console.log(this.fooString);
  }

  doBar() {
    return this.barSize /2;
  }

  doubleBar() {
    return this.barSize *2;
  }
}

class Baz {
  foo = new Foo();

  doBaz() {
    this.foo.doBar();
    this.foo.doFoo();
  }
}
```

Executing "Extract Class" on `Foo` will pop up this modal to ask for the things you want to extract:

![image](https://user-images.githubusercontent.com/1094774/96325826-44c58200-0ff9-11eb-96f4-891b4a5bd907.png)

And will generate such code:

```ts
class Foo {
  private extractedClass: ExtractedClass = new ExtractedClass();

  get barSize(): number {
    return this.extractedClass.barSize;
  }

  fooString: string = "baz";

  doFoo() {
    console.log(this.fooString);
  }

  doBar() {
    return this.extractedClass.doBar();
  }

  doubleBar() {
    return this.extractedClass.doubleBar();
  }
}

class ExtractedClass {
  barSize: number = 30;

  doBar() {
    return this.barSize / 2;
  }

  doubleBar() {
    return this.barSize * 2;
  }
}

class Baz {
  foo = new Foo();

  doBaz() {
    this.foo.doBar();
    this.foo.doFoo();
  }
}
```

Note that `Baz` hasn't change. To preserve the existing behavior, everything was preserved on `Foo`. So the next steps for you is to progressively adapt the clients to use the new class instead, so you can finally get rid of the things in `Foo` and effectively complete the refactoring.

The good thing is that all the part of creating a new class and moving the stuff is done automatically, without breaking the code.

The `ExtractedClass` is generated in the same file. We might enhance that later though, door is open to iterations.

**Comment by nicoespeon** [comment]
This will be part of the next release. I'll do that very soon (just a few things to document before doing so, but you can expect a release by next week).

**Comment by nicoespeon** [comment]
FYI, this is now available with the latest release (v4.9.0) published today: https://marketplace.visualstudio.com/items?itemName=nicoespeon.abracadabra#extract-class

**Comment by avik-so** [comment]
Amazing! Thank you so much!!!!
